packages/devextreme/js/__internal/core/utils/m_date.ts

```diff
@@ -222,6 +222,21 @@ const getDatesDifferences = function (date1, date2) {
   return differences;
 };
 
+function getMonthInterval(value, dir) {
+  const result = new Date(value.getTime());
+  const originalDate = result.getDate();
+  const targetMonth = result.getMonth() + dir;
+
+  result.setDate(1);
+  result.setMonth(targetMonth);
+
+  const lastDayOfTargetMonth = new Date(result.getFullYear(), result.getMonth() + 1, 0).getDate();
+
+  result.setDate(Math.min(originalDate, lastDayOfTargetMonth));
```

Comment:
[imperative, non-intuitive]
The implementation is difficult enough, I cannot say it works correctly for all cases without significant testing.
I see that in calendar you actually don't need to shift months you need only a first or last date of the update month. Therefore, I suggest you to just update `getFirstMonthDate` and `getLastMonthDate` utils with a new parameter (delta) to achieve your goal.